kotlin-runtime/src/main/kotlin/io/micronaut/context/env/hocon/HoconPropertySourceLoader.kt

```diff
@@ -36,22 +36,57 @@ import kotlin.collections.LinkedHashMap
  * A PropertySourceLoader that supports config4k.
  *
  * @author graemerocher
+ * @author Marcelo Liberato
  * @since 1.1
  */
 class HoconPropertySourceLoader : PropertySourceLoader, Ordered {
 
+    private val impl by lazy {
+        ClassUtils
+                .forName("io.micronaut.context.env.hocon.HoconPropertySourceLoaderImpl", HoconPropertySourceLoader::class.java.classLoader)
+                .get()
+                .newInstance() as PropertySourceLoader
+    }
+
     override fun getExtensions(): MutableSet<String> {
         return Collections.singleton("conf")
     }
-    
-    override fun load(resourceName: String?, resourceLoader: ResourceLoader?): Optional<PropertySource> {
-        return loadEnv(resourceName, resourceLoader, null)
-    }
 
     override fun isEnabled(): Boolean {
         return ClassUtils.isPresent("com.typesafe.config.Config", HoconPropertySourceLoader::class.java.classLoader)
     }
 
+    override fun load(resourceName: String?, resourceLoader: ResourceLoader?): Optional<PropertySource> {
+        return if(isEnabled) {
+            impl.load(resourceName, resourceLoader)
+        } else {
+            Optional.empty()
```

Comment:
If this else block is ever reached, I'm not sure falling back to an empty value is a good idea. Users can get confused when the file exists but properties are still not found, and it will be pretty hard to know that it's because the feature is not enabled.
I think an exception would be better.

Reply:
in theory it should never be reached when executing under normal circumstances

Reply:
It should never get there, BUT, "load" is been called by the framework even if "isEnabled" is false. Had to do that to prevent creating Impl class when there is no config4k in classpath